File: src/leetcode/right_side_view.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
struct Solution {}
use std::{cell::RefCell, collections::VecDeque, rc::Rc};

impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut view = Vec::new();
        let mut queue = VecDeque::new();
        if let Some(r) = root {
            queue.push_back(r.clone());
        }
        while !queue.is_empty() {
            let level = queue.len();
            view.push(queue.front().unwrap().borrow().val);
            for _ in 0..level {
                let n = queue.pop_front().unwrap();
                if let Some(r) = &n.borrow().right {
                    queue.push_back(r.clone());
                };
                if let Some(l) = &n.borrow().left {
                    queue.push_back(l.clone());
                };
            }
        }
        view
    }
}

File: leetcode_cpp/Leetcode75/BinaryTree/rightSideView.cpp
#include <bits/stdc++.h>
using namespace std;
auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode():
      val(0),
      left(nullptr),
      right(nullptr) {}

    explicit TreeNode(int x):
      val(x),
      left(nullptr),
      right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right):
      val(x),
      left(left),
      right(right) {}
};

class Solution {
  public:
    static vector<int> rightSideView(TreeNode *root) {
        std::vector<int> ans;
        if (!root) {
            return ans;
        }
        std::vector<TreeNode *> v{root};
        std::vector<TreeNode *> tmp;
        while (!v.empty()) {
            ans.push_back(v.back()->val);
            for (auto node : v) {
                if (node->left) {
                    tmp.push_back(node->left);
                }
                if (node->right) {
                    tmp.push_back(node->right);
                }
            }
            v.swap(tmp);
            tmp.clear();
        }
        return ans;
    }
};